apps/arweave/src/ar_mining_worker.erl

```diff
@@ -245,281 +277,181 @@ terminate(_Reason, _State) ->
-			{chunk, WhichChunk}]),
-	State.
+-spec handle_task(
+	Task :: {
+		EventType :: compute_h0 | computed_h0 | chunk1 | chunk2 | computed_h1 | computed_h2 | compute_h2_for_peer,
+		Candidate :: #mining_candidate{},
+		ExtraArgs :: term()
+	},
+	State :: #state{}
+) -> {noreply, State :: #state{}}.
 
-%% @doc Return the sub-chunk and the remaining bytes in the chunk. For spora_2_6 packing
-%% (aka difficulty 0), each sub-chunk is the size of a chunk so there are no remaining
-%% bytes.
-extract_sub_chunk(Chunk, #mining_candidate{ packing_difficulty = 0 }) ->
-	{Chunk, <<>>};
-extract_sub_chunk(Chunk, _Candidate) ->
-	<< SubChunk:?COMPOSITE_PACKING_SUB_CHUNK_SIZE/binary, Rest/binary >> = Chunk,
-	{SubChunk, Rest}.
-
-process_sub_chunk(chunk1, Candidate, SubChunk, State) ->
-	ar_mining_hash:compute_h1(self(), Candidate#mining_candidate{ chunk1 = SubChunk }),
-	State;
-process_sub_chunk(chunk2, Candidate, SubChunk, State) ->
-	#mining_candidate{ session_key = SessionKey } = Candidate,
-	Candidate2 = Candidate#mining_candidate{ chunk2 = SubChunk },
-	case cycle_sub_chunk_cache(Candidate2, {chunk2, SubChunk}, State) of
-		{{chunk1, Chunk1, H1}, State2} ->
-			ar_mining_hash:compute_h2(
-				self(), Candidate2#mining_candidate{ chunk1 = Chunk1, h1 = H1 }),
-			%% Decrement 2 for chunk1 and chunk2:
-			%% 1. chunk1 was previously read and cached
-			%% 2. chunk2 that was just read and will shortly be used to compute h2
-			update_sub_chunk_cache_size(-2, SessionKey, State2);
-		{{chunk1, H1}, State2} ->
-			ar_mining_hash:compute_h2(self(), Candidate2#mining_candidate{ h1 = H1 }),
-			%% Decrement 1 for chunk2:
-			%% we're computing h2 for a peer so chunk1 was not previously read or cached 
-			%% on this node
-			update_sub_chunk_cache_size(-1, SessionKey, State2);
-		{do_not_cache, State2} ->
-			%% Decrement 1 for chunk2
-			%% do_not_cache indicates chunk1 was not and will not be read or cached
-			update_sub_chunk_cache_size(-1, SessionKey, State2);
-		{cached, State2} ->
-			case Candidate2#mining_candidate.cm_lead_peer of
-				not_set ->
-					ok;
-				_ ->
-					?LOG_ERROR([{event, cm_chunk2_cached_before_chunk1},
-						{worker, State#state.name},
-						{partition_number, Candidate2#mining_candidate.partition_number},
-						{partition_number2, Candidate2#mining_candidate.partition_number2},
-						{cm_peer, ar_util:format_peer(Candidate2#mining_candidate.cm_lead_peer)},
-						{cache_ref, Candidate2#mining_candidate.cache_ref},
-						{nonce, Candidate2#mining_candidate.nonce},
-						{session, ar_nonce_limiter:encode_session_key(SessionKey)}])
-			end,
-			State2
-	end.
+%% @doc Handle the `compute_h0` task.
+%% Indicates that the VDF step has been computed.
+handle_task({compute_h0, Candidate, _ExtraArgs}, State) ->
+	#state{
+		latest_vdf_step_number = LatestVDFStepNumber,
+		vdf_queue_limit = VDFQueueLimit
+	} = State,
+	#mining_candidate{ step_number = StepNumber } = Candidate,
+	State2 = report_and_reset_hashes(State),
```

Comment:
I forget, why did we decide to move the `try_to_reserve_cache_space` out of `compute_h0` and into `computed_h0`?

It does open the possibility of the node computing a bunch of h0s that it then has to throw away (i.e. computes them but then can't process them further cause th cache is full, but by the time the cache has space the previously computed h0s are now too old to use).

I'm not sure how big an impact it will have in practice - so if there's a good reason to delay the cache check until later it's probably fine. But if it's all the same, checking the cache earlier should reduce some amount of unnecessary work. And computing h0 isn't *expensive* but it's not cheap (it's a randomx operation)

Reply:
Okay yeah I think we do have to do the cache check here as this is the only place we check `StepNumber >= LatestVDFStepNumber - VDFQueueLimit`

By the time we make it to `computed_h0` we assume we're going to proceed through the whole pipeline, which means we may end up reading chunks and performing hashes for a too-old step number.

Best is to have the cache check, VDFQueueLimit check, and the `add_delayed_task` line here, in `compute_h0`. That way by the time we reach computed_h0 we know the step number is fresh and the cache has space.